File: CollectionView_WaterfallFlow/CollectionView_WaterfallFlow/ViewController.m
//
//  ViewController.m
//  CollectionView_WaterfallFlow
//
//  Created by 陈博 on 2018/3/8.
//  Copyright © 2018年 陈博. All rights reserved.
//

#import "ViewController.h"
#import "WaterCollectionViewCell.h"



@interface ViewController ()<UICollectionViewDelegate,UICollectionViewDataSource>
@property (nonatomic, strong) UICollectionView * demoCollectionView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UICollectionViewFlowLayout * layout = [[UICollectionViewFlowLayout alloc] init];
    layout.itemSize = CGSizeMake((self.view.frame.size.width - 15) / 2, 300);
    
    _demoCollectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 20, self.view.frame.size.width, self.view.frame.size.height) collectionViewLayout:layout];
    [_demoCollectionView registerClass:[WaterCollectionViewCell class] forCellWithReuseIdentifier:@"WaterCollectionViewCell"];
    _demoCollectionView.backgroundColor = [UIColor grayColor];
    _demoCollectionView.delegate = self;
    _demoCollectionView.dataSource = self;
    [self.view addSubview:_demoCollectionView];
    
}
#pragma mark - UICollectionViewDelegate&&UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return 6;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    UICollectionViewCell * cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"WaterCollectionViewCell" forIndexPath:indexPath];
    if (cell == nil) {
        cell = [[WaterCollectionViewCell alloc] initWithFrame:CGRectMake(0, 0, (self.view.frame.size.width - 15) / 2, 300)];
        
    }
    return cell;
}

@end

File: Swift-瀑布流/Swift-瀑布流/ViewController.swift
//
//  ViewController.swift
//  Swift-瀑布流
//
//  Created by 陈博 on 2018/4/24.
//  Copyright © 2018年 陈博. All rights reserved.
//

import UIKit
private let kContentCellID = "kContentCellID"
class ViewController: UIViewController {
    fileprivate lazy var collectionView : UICollectionView = {
       let layout = WaterFallLayout()
        
        // 设置列数
        let cols : CGFloat = 3
        // 设置item之间的间距
        let itemMargin : CGFloat = 8
        let itemWH = (self.view.bounds.width - (cols + 1) * itemMargin) / cols
        layout.itemSize = CGSize(width: itemWH, height: itemWH)
        layout.minimumLineSpacing = itemMargin
        layout.minimumInteritemSpacing = itemMargin
        
        
        // 设置内边距
        layout.sectionInset = UIEdgeInsets(top: itemMargin, left: itemMargin, bottom: itemMargin, right: itemMargin)
        
       let collectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: layout)
        collectionView.backgroundColor = UIColor.white
        collectionView.dataSource = self
        
        return collectionView
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white
        view.addSubview(collectionView)
        collectionView .register(UICollectionViewCell.self, forCellWithReuseIdentifier: kContentCellID)
        
    }

}
extension ViewController: UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 100
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: kContentCellID, for: indexPath)
        cell.backgroundColor = UIColor.green
        return cell
    }
}

